src/ProducerEventProcessor/ProducerEventProcessorTest.cpp

```diff
@@ -1321,33 +1324,189 @@ TEST(ProducerEventProcessor, ThreadNameSmoke) {
+    const ThreadStateSlice& expected_thread_state_slice) {
+  return ::testing::Property("thread_state_slice", &ClientCaptureEvent::thread_state_slice,
+                             ThreadStateSliceEq(expected_thread_state_slice));
+}
+
+::testing::Matcher<ClientCaptureEvent> ClientCaptureEventsInternedCallstackEq(
+    const InternedCallstack& expected_interned_callstack) {
+  return ::testing::Property("interned_callstack", &ClientCaptureEvent::interned_callstack,
+                             InternedCallstackEq(expected_interned_callstack));
+}
+
 TEST(ProducerEventProcessor, ThreadStateSliceSmoke) {
   MockClientCaptureEventCollector collector;
   auto producer_event_processor = ProducerEventProcessor::Create(&collector);
 
   ProducerCaptureEvent producer_event;
-  {
-    ThreadStateSlice* thread_state_slice = producer_event.mutable_thread_state_slice();
-    thread_state_slice->set_pid(kPid1);
-    thread_state_slice->set_tid(kTid1);
-    thread_state_slice->set_thread_state(ThreadStateSlice::kIdle);
-    thread_state_slice->set_duration_ns(kDurationNs1);
-    thread_state_slice->set_end_timestamp_ns(kTimestampNs1);
-  }
+  ThreadStateSlice* producer_thread_state_slice = producer_event.mutable_thread_state_slice();
+  producer_thread_state_slice->set_pid(kPid1);
+  producer_thread_state_slice->set_tid(kTid1);
+  producer_thread_state_slice->set_thread_state(ThreadStateSlice::kIdle);
+  producer_thread_state_slice->set_duration_ns(kDurationNs1);
+  producer_thread_state_slice->set_end_timestamp_ns(kTimestampNs1);
+  producer_thread_state_slice->set_switch_out_or_wakeup_callstack_status(
+      ThreadStateSlice::kNoCallstack);
+  producer_thread_state_slice->set_switch_out_or_wakeup_callstack_id(0);
+
+  ThreadStateSlice expected_thread_state_slice = *producer_thread_state_slice;
 
   ClientCaptureEvent event;
-
   EXPECT_CALL(collector, AddEvent).Times(1).WillOnce(SaveArg<0>(&event));
 
   producer_event_processor->ProcessEvent(1, std::move(producer_event));
 
   ASSERT_EQ(event.event_case(), ClientCaptureEvent::kThreadStateSlice);
-  const ThreadStateSlice& thread_state_slice = event.thread_state_slice();
-  EXPECT_EQ(thread_state_slice.pid(), kPid1);
-  EXPECT_EQ(thread_state_slice.tid(), kTid1);
-  EXPECT_EQ(thread_state_slice.thread_state(), ThreadStateSlice::kIdle);
-  EXPECT_EQ(thread_state_slice.duration_ns(), kDurationNs1);
-  EXPECT_EQ(thread_state_slice.end_timestamp_ns(), kTimestampNs1);
+  EXPECT_THAT(event, ClientCaptureEventsTheadStateSliceEq(expected_thread_state_slice));
+}
+
+TEST(ProducerEventProcessor, ThreadStateSliceMergesCallstack) {
+  MockClientCaptureEventCollector collector;
+  auto producer_event_processor = ProducerEventProcessor::Create(&collector);
+
+  constexpr uint64_t kCallstackFrame1 = 1;
+  constexpr uint64_t kCallstackFrame2 = 2;
+  constexpr uint64_t kCallstackFrame3 = 3;
+
+  ProducerCaptureEvent thread_state_slice_callstack_producer_event1;
```

Comment:
I don't think you need the `_producer_event` suffix in all these names. It makes it a bit harder to read.

Reply:
Agree. Still it needs to be distinguished from the content. I went with `_event` as a suffix.